specs/rollup-node-p2p.md

```diff
@@ -0,0 +1,258 @@
+
+##### Message ID computation
+
+[Same as L1][l1-message-id], with recognition of compression:
+
+- If `message.data` has a valid snappy decompression, set `message-id` to the first 20 bytes of the `SHA256` hash of
+  the concatenation of `MESSAGE_DOMAIN_VALID_SNAPPY` with the snappy decompressed message data,
+  i.e. `SHA256(MESSAGE_DOMAIN_VALID_SNAPPY + snappy_decompress(message.data))[:20]`.
+- Otherwise, set `message-id` to the first 20 bytes of the `SHA256` hash of
+  the concatenation of `MESSAGE_DOMAIN_INVALID_SNAPPY` with the raw message data,
+  i.e. `SHA256(MESSAGE_DOMAIN_INVALID_SNAPPY + message.data)[:20]`.
+
+#### Heartbeat and parameters
+
+GossipSub [parameters][gossip-parameters]:
+
+- `D` (topic stable mesh target count): 8
+- `D_low` (topic stable mesh low watermark): 6
+- `D_high` (topic stable mesh high watermark): 12
+- `D_lazy` (gossip target): 6
+- `heartbeat_interval` (frequency of heartbeat, seconds): 0.5
+- `fanout_ttl` (ttl for fanout maps for topics we are not subscribed to but have published to, seconds): 24
+- `mcache_len` (number of windows to retain full messages in cache for `IWANT` responses): 12
+- `mcache_gossip` (number of windows to gossip about): 3
+- `seen_ttl` (number of heartbeat intervals to retain message IDs): 40
+
+Notable differences from L1 consensus (Eth2):
+
+- `seen_ttl` does not need to cover a full L1 epoch (6.4 minutes), but rather just a small window covering latest blocks
+- `fanout_ttl`: adjusted to lower than `seen_ttl`
+- `mcache_len`: a larger number of heartbeats can be retained since the gossip is much less noisy.
+- `heartbeat_interval`: faster interval to reduce latency, bandwidth should still be reasonable since
+  there are far fewer messages to gossip about each interval than on L1 which uses an interval of 0.7 seconds.
+
+#### Topic configuration
+
+Topics have string identifiers and are communicated with messages and subscriptions.
+`/optimism/chain_id/hardfork_version/Name`
+
+- `chain_id`: replace with decimal representation of chain ID
+- `hardfork_version`: replace with decimal representation of hardfork
+- `Name`: topic application-name
+
+Note that the topic encoding depends on the topic, unlike L1,
+since there are less topics, and all are snappy-compressed.
+
+## Gossip Topics
+
+### `blocks`
+
+The primary topic of the L2, to distribute blocks to other nodes faster than proxying through L1 would.
+
+#### Block encoding
+
+TODO: encode execution payload (SSZ or RLP), with sequencer identifier and signature.
+
+TODO: signature type and verification (`secp256k1` like transactions, with some different signature domain)
+
+#### Block validation
+
+To ensure malicious peers get scored based on application behavior the validation signals
+`ACCEPT` (valid), `IGNORE` (like inactivity) or `REJECT` (score penalties).
+
+In order of operation:
+
+- `[REJECT]` if the encoding or compression is not valid
+- `[REJECT]` if the block timestamp is older than 20 seconds in the past
```

Comment:
If the block timestamp is older than 20 seconds in the past, how would it be possible to sync historically? This can only apply if we are at the tip, but how does one know if they are at the tip?

Reply:
Historic block-by-block sync wouldn't happen through gossipsub. Instead the peer should just read the data from L1. And if we want p2p historic block-by-block sync (if it's even faster than pulling the data from L1) then we can specify a req-resp protocol in libp2p (libp2p enables you to open named read/write channels with peers, if the peers negotiate the support for that channel protocol during their connection).

For long-range happy-case sync though, we only need *any recent* header. We verify that it matches the output root on L1 if we want more of a guarantee we're syncing a correct chain (otherwise have to reorg it all out again if it's wrong), and then call the engine with a new-payload call that doesn't simply extend the chain by 1 block, but is that recent header. The engine will then start snap-sync (or other type of sync) towards that header, on its own. This is a later implementation step to trigger, we just need recent blocks for now.

Reply:
> Instead the peer should just read the data from L1

If blob transactions are a thing, then this will not be possible. It also sounds like you are really in favor of an eip4444 like mechanism. I agree that we need to be able to snap sync via a single recent header that is greater than 1 week old (finality, knowing we are on the correct chain). I do think that it is important to be able to spin up archive nodes though which requires historical execution because no existing sync mechanism is designed specifically for archive nodes